From: Qu Wenruo <wqu@suse.com>
To: linux-btrfs@vger.kernel.org
Subject: btrfs: tree-checker: Add DATA_RELOC tree to essential non-empty tree list
Date: Thu, 19 Jul 2018 15:05:14 +0800

DATA RELOC tree should also never be empty.

[HH: backported to 4.16.x]
Link: https://bugzilla.kernel.org/show_bug.cgi?id=200407
Reported-by: Xu Wen <wen.xu@gatech.edu>
Signed-off-by: Qu Wenruo <wqu@suse.com>
Reviewed-by: Su Yue <suy.fnst@cn.fujitsu.com>
---
It's possible just fold this fix into commit "btrfs: tree-checker: Detect
invalid empty essential tree".

And for the test image, it can already be rejected by misc-next branch
by the enhanced dev-extent checker.
---
 fs/btrfs/tree-checker.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/fs/btrfs/tree-checker.c b/fs/btrfs/tree-checker.c
index 4c82a628df85..db835635372f 100644
--- a/fs/btrfs/tree-checker.c
+++ b/fs/btrfs/tree-checker.c
@@ -504,9 +504,10 @@ static int check_leaf(struct btrfs_fs_info *fs_info, struct extent_buffer *leaf,
 		    owner == BTRFS_CHUNK_TREE_OBJECTID ||
 		    owner == BTRFS_EXTENT_TREE_OBJECTID ||
 		    owner == BTRFS_DEV_TREE_OBJECTID ||
-		    owner == BTRFS_FS_TREE_OBJECTID) {
+		    owner == BTRFS_FS_TREE_OBJECTID ||
+		    owner == BTRFS_DATA_RELOC_TREE_OBJECTID) {
 			generic_err(root, leaf, 0,
-			"invalid root, root %llu should never be empty",
+			"invalid root, root %llu must never be empty",
 				    owner);
 			return -EUCLEAN;
 		}
